mixnet/client/src/sender.rs

```diff
@@ -133,21 +133,58 @@ where
             return None;
         }
 
+        self.remaining_delays -= 1;
+
+        // corner case: 0 average delay
+        if self.avg_delay == 0 {
+            return Some(0);
+        }
+
         if self.remaining_delays == 1 {
-            self.remaining_delays -= 1;
             return Some(self.remaining_time);
         }
 
+        // if we have no remaining time, then just return 0
+        if self.remaining_time == 0 {
+            return Some(0);
+        }
+
         // Calculate bounds to avoid extreme values
-        let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0);
```

Comment:
`.min((self.remaining_time as f64).saturating_sub(self.remaining_delays - 1))` and I think we can remove both the avg delay and remaining time edge cases

Reply:
If we remove the avg delay, then we may have extreme delays, and with higher possibility delay1 > delay 2 > delay3 and etc.

Reply:
nono, I mean just removing the edge case at line 139

Reply:
Ah, yeah, sorry for the misunderstanding.

Reply:
I still think we need the saturating sub in `.min((self.remaining_time as f64).saturating_sub(self.remaining_delays - 1))`